eureka2-client/src/test/java/com/netflix/eureka2/client/channel/RetryableConnectionFactoryTest.java

```diff
@@ -0,0 +1,468 @@
+            this.id = id;
+        }
+
+        @Override
+        public String toString() {
+            return "Op:" + id;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof TestOp)) return false;
+
+            TestOp testOp = (TestOp) o;
+
+            if (id != testOp.id) return false;
+
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            return id;
+        }
+    }
+
+    static class WorkAndResult {
+        public enum Result {onCompleted, onError}
+
+        public final TestOp op;
+        public final Result result;
+
+        public WorkAndResult(TestOp op, Result result) {
+            this.op = op;
+            this.result = result;
+        }
+
+        @Override
+        public String toString() {
+            return "WorkAndResult{" +
+                    "op=" + op +
+                    ", result=" + result +
+                    '}';
+        }
+    }
+
+    static class TestServiceChannel implements ServiceChannel {
+        public final int id;
+        public final Func1<TestOp, Boolean> failurePredicate;
+
+        public final ConcurrentLinkedQueue<WorkAndResult> results = new ConcurrentLinkedQueue<>();
+        public final ReplaySubject<Void> channelLifecycle = ReplaySubject.create();
+        public final AtomicBoolean closed = new AtomicBoolean(false);
+
+        public TestServiceChannel(int id, Func1<TestOp, Boolean> failurePredicate) {
+            this.id = id;
+            this.failurePredicate = failurePredicate;
+        }
+
+        public Observable<Void> doWork(TestOp op) {
+            if (closed.get()) {
+                return Observable.error(new Exception("already closed"));
+            }
+
+            if (failurePredicate.call(op)) {
+                results.add(new WorkAndResult(op, WorkAndResult.Result.onError));
+                Exception e = new Exception("error triggered");
+                close(e);
+                return Observable.error(e);
+            } else {
+                results.add(new WorkAndResult(op, WorkAndResult.Result.onCompleted));
+                return Observable.empty();
+            }
+        }
+
+        @Override
+        public void close() {
+            if (closed.compareAndSet(false, true)) {
+                channelLifecycle.onCompleted();
+            }
+        }
+
+        @Override
+        public void close(Throwable error) {
+            if (closed.compareAndSet(false, true)) {
+                channelLifecycle.onError(error);
+            }
+        }
+
+        @Override
+        public Observable<Void> asLifecycleObservable() {
+            return channelLifecycle.asObservable();
+        }
+
+        @Override
+        public String toString() {
+            return "channel:" + id;
+        }
+    }
+
+    static class TestChannelFactory implements ChannelFactory<TestServiceChannel> {
```

Comment:
Ahh yes. This TestChannelFactory was created before that one. Indeed they can be merged. Will do that in a separate pull.